onnxruntime/core/optimizer/nhwc_transformer.cc

```diff
@@ -14,48 +14,56 @@ using namespace onnx_layout_transformation;
-  ORT_UNUSED_PARAMETER(graph);
-
-  // Skip if domain is incorrect
-  auto domain = node.Domain();
-  if (domain != "" && domain != "com.microsoft") {
-    return {false, 0, std::nullopt, std::nullopt};
-  }
-
-  // Skip if already transformed
-  if (node.GetAttributeIntDefault("channels_last", 0) == 1) {
-    return {false, 0, std::nullopt, std::nullopt};
-  }
-
-  // Skip if unknown rank
-  auto shape = NodeFromApiNode(node).InputDefs()[0]->Shape();
-  if (shape == nullptr) {
-    return {false, 0, std::nullopt, std::nullopt};
-  }
-
-  // Convert to channels last
-  size_t rank = shape->dim_size();
-  node.SetAttributeInt("channels_last", 1);
-  return {true, rank, std::nullopt, "com.microsoft"};
-}
-
 Status NhwcTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
   GraphViewer graph_viewer(graph);
   for (auto index : graph_viewer.GetNodesInTopologicalOrder()) {
     auto& node = *graph.GetNode(index);
     ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
   }
 
-  // Only QLinearConv needs to be handled explicitly. The rest will be transformed if needed during transpose
-  // optimization.
-  std::unordered_map<std::string_view, LayoutHandler> handler_map = {
-      {"QLinearConv", &QLinearConvHandler}
-  };
+  auto api_graph = MakeApiGraph(graph, cpu_allocator_, logger, kCpuExecutionProvider);
+
+  modified = false;
+  for (std::unique_ptr<api::NodeRef>& node : api_graph->Nodes()) {
+    // Only QLinearConv needs to be handled explicitly. The rest will be transformed if needed during transpose
+    // optimization.
+    if (node->OpType() == "QLinearConv") {
+      auto domain = node->Domain();
+
+      // Skip if domain is incorrect
+      if (domain != kOnnxDomain && domain != kOnnxDomainAlias && domain != kMSDomain) {
+        continue;
+      }
+
+      // Skip if already transformed
+      if (node->GetAttributeIntDefault("channels_last", 0) == 1) {
+        continue;
+      }
+
+      // Skip if unknown rank
+      auto shape = NodeFromApiNode(*node).InputDefs()[0]->Shape();
+      if (shape == nullptr) {
+        continue;
+      }
+
+      // Convert to channels last
+      size_t rank = shape->dim_size();
+      node->SetAttributeInt("channels_last", 1);
+
+      std::vector<int64_t> input_perm = ChannelFirstToLastPerm(rank);
+      std::vector<int64_t> output_perm = ChannelLastToFirstPerm(rank);
+      WrapTransposesAroundNode(*api_graph, *node, {&input_perm}, {&output_perm});
+
+      if (domain != kMSDomain) {
+        SwapNodeOpTypeAndDomain(*api_graph, *node, "QLinearConv", "com.microsoft");
+      }
+
+      modified = true;
+    }
+  }
 
-  auto api_graph = MakeApiGraph(graph, std::move(cpu_allocator_), logger, kCpuExecutionProvider);
-  if (ChannelFirstToChannelLast(*api_graph, handler_map, /*allow_extended_ops*/ true)) {
-    modified = true;
+  if (modified) {
+    Optimize(*api_graph, /*allow_extended_ops*/ true);
   }
```

Comment:
If Optimize recurses does it only need to be called when level is 1? 

We do that with Graph::Resolve, but it's slightly different in that it must always start at the main graph even if called from a subgraph.

Reply:
Optimize only optimizes the current graph level.